stl/inc/span

```diff
@@ -0,0 +1,690 @@
+
+_STD_BEGIN
+
+inline constexpr size_t dynamic_extent = static_cast<size_t>(-1);
+
+// STRUCT TEMPLATE _Span_iterator
+template <class _Ty>
+struct _Span_iterator {
+#ifdef __cpp_lib_concepts
+    using iterator_concept = contiguous_iterator_tag;
+#endif // __cpp_lib_concepts
+    using iterator_category = random_access_iterator_tag;
+    using value_type        = remove_cv_t<_Ty>;
+    using difference_type   = ptrdiff_t;
+    using pointer           = _Ty*;
+    using reference         = _Ty&;
+
+    constexpr _Span_iterator() noexcept = default;
+
+#if _ITERATOR_DEBUG_LEVEL >= 1
+    constexpr _Span_iterator(const pointer _Pointer, const pointer _Begin, const pointer _End) noexcept
+        : _Mybegin(_Begin), _Myend(_End), _Myptr(_Pointer) {}
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+    constexpr explicit _Span_iterator(const pointer _Pointer) noexcept : _Myptr(_Pointer) {}
+#endif // _ITERATOR_DEBUG_LEVEL
+
+    _NODISCARD constexpr reference operator*() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(_Mybegin, "cannot dereference value-initialized span iterator");
+        _STL_VERIFY(_Myptr < _Myend, "cannot dereference end span iterator");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        return *_Myptr;
+    }
+
+    _NODISCARD constexpr pointer operator->() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(_Mybegin, "cannot dereference value-initialized span iterator");
+        _STL_VERIFY(_Myptr < _Myend, "cannot dereference end span iterator");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        return _Myptr;
+    }
+
+    constexpr _Span_iterator& operator++() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(_Mybegin, "cannot increment value-initialized span iterator");
+        _STL_VERIFY(_Myptr < _Myend, "cannot increment span iterator past end");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        ++_Myptr;
+        return *this;
+    }
+
+    constexpr _Span_iterator operator++(int) noexcept {
+        _Span_iterator _Tmp{*this};
+        ++*this;
+        return _Tmp;
+    }
+
+    constexpr _Span_iterator& operator--() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        _STL_VERIFY(_Mybegin, "cannot decrement value-initialized span iterator");
+        _STL_VERIFY(_Mybegin < _Myptr, "cannot decrement span iterator before begin");
+#endif // _ITERATOR_DEBUG_LEVEL >= 1
+        --_Myptr;
+        return *this;
+    }
+
+    constexpr _Span_iterator operator--(int) noexcept {
+        _Span_iterator _Tmp{*this};
+        --*this;
+        return _Tmp;
+    }
+
+    constexpr void _Verify_offset([[maybe_unused]] const difference_type _Off) const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        if (_Off != 0) {
+            _STL_VERIFY(_Mybegin, "cannot seek value-initialized span iterator");
+        }
+
+        if (_Off < 0) {
+#pragma warning(suppress : 4146) // unary minus operator applied to unsigned type, result still unsigned

```

Comment:
Is this suppression really needed, or an artifact of earlier implementation? I see `-_Off` on the next line, but `_Off` has a signed type. (Also on line 137.)

Reply:
I think the `-_Off` stems from writing the canonical `_Myptr - _Mybegin` rather than `_Mybegin - _Myptr ` as we are dealing with decrements.

But you are correct the warning should not be needed anymore